### PR TITLE
add upload node

### PR DESCRIPTION
**Pull request by karo-fox**

### CODE REVIEW

File: src/upload_node.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        self.heading = None
+        self.rel_altitude = None
+
+
+class UploadNode:
+    def __init__(self):
+        self.global_pose_sub = rospy.Subscriber("uav0/mavros/global_position/global", NavSatFix, self.global_pose_callback)
+        self.heading_sub = rospy.Subscriber("uav0/mavros/global_position/compass_hdg", Float64, self.heading_callback)
+        self.alt_sub = rospy.Subscriber("uav0/mavros/global_position/rel_alt", Float64, self.altitude_callback)
+        self.figure_sub = rospy.Subscriber("uav0/detection/confirmed_figures", FigureMsgList, self.figure_callback)
```
Comment (RocketFan) [inline src/upload_node.py:33]: uav0 prefix will be added in roslaunch so you can remove it

File: src/upload_node.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            if self.ready:
+                if self.drone_data.is_complete():
+                    data = self.get_drone_request_data()
+                    self.drone_uploader.add(data)
+                    self.drone_data.reset()
+                if self.figures is not None:
+                    for fig in self.figures:
+                        data = self.get_figure_request_data(fig)
+                        self.figure_uploader.add(data)
+                    self.figures = None
```
Comment (RocketFan) [inline src/upload_node.py:53]: Move to figure callback

File: src/upload_node.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        self.start_pose = DroneData()
+        self.drone_uploader = Uploader("http://91.227.41.24:3001/drones")
+        self.figure_uploader = Uploader("http://91.227.41.24:3001/samples")
+        self.drone_data = DroneData()
+        self.rate = rospy.Rate(10)
+        self.ready = False
+        self.figures = None
+
+    def run(self):
+        while True:
```
Comment (RocketFan) [inline src/upload_node.py:43]: Use `not rospy.is_shutdown()`  instead of while True

File: src/upload_node.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                self.start_pose.global_pose.longitude,
+                self.start_pose.heading.data
+            )
+
+            self.ready = True
+
+    def global_pose_callback(self, data):
+        if self.start_pose.global_pose is None:
+            self.start_pose.global_pose = data
+            return
```
Comment (RocketFan) [inline src/upload_node.py:72]: Use `rospy.wait_for_message` to initialize variables

File: src/upload_node.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        self.drone_data.global_pose = data
+
+    def heading_callback(self, data):
+        if self.start_pose.heading is None:
+            self.start_pose.heading = data
+            return
```
Comment (RocketFan) [inline src/upload_node.py:78]: Use rospy.wait_for_message to initialize variables

File: src/upload_node.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                rospy.loginfo("Waiting for start_pose")
+                continue
+
+            self.coordinate_scaler = CoordinateScaler(
+                self.start_pose.global_pose.latitude,
+                self.start_pose.global_pose.longitude,
+                self.start_pose.heading.data
+            )
+
+            self.ready = True
```
Comment (RocketFan) [inline src/upload_node.py:67]: If you will use `rospy.wait_for_message` for start_pose, you can initialize it in constructor

File: src/upload_node.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    heading: Float64 = None
+    rel_altitude: Float64 = None
+
+    def is_complete(self) -> bool:
+        return all(param is not None for param in (self.global_pose, self.heading, self.rel_altitude))
+
+    def reset(self) -> None:
+        self.global_pose = None
+        self.heading = None
+        self.rel_altitude = None
```
Comment (RocketFan) [inline src/upload_node.py:25]: Use [message_filters](http://wiki.ros.org/message_filters) to get one callback from multiple subscribers, you can see example at figure finder node

File: src/upload_node.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        if figure.type == "purpleBall":
+            return "p"
+        if figure.type == "blueBall":
+            return "b"
+
+
+if __name__ == "__main__":
+    rospy.init_node("upload_node_py")
+    upload_node = UploadNode()
+    upload_node.run()
```
Comment (RocketFan) [inline src/upload_node.py:117]: Wrap it inside `try: ... catch rospy.ROSInterruptException: pass`

**Review by RocketFan — CHANGES_REQUESTED** [review]

File: src/upload_node.py (hunk 3 above)
Comment (karo-fox) [reply src/upload_node.py]: not needed anymore, everything has moved to callbacks / __init__()

File: src/upload_node.py (hunk 7 above)
Comment (karo-fox) [reply src/upload_node.py]: I used ApproximateTimeSynchronizser, because we get Float64 messages which are headerless

**Review by RocketFan — APPROVED** [review]

**Pull request merged**